File: dotnet/return-struct-from-callbacks/src/ReturnStructFromCallbacks/ReturnStructFromCallbacksStack.cs
using System;
using Amazon.CDK;
using Events = Amazon.CDK.AWS.Events;
using Amazon.CDK.AWS.EC2;
using Amazon.CDK.AWS.ElasticLoadBalancingV2;

namespace ReturnStructFromCallbacks
{
    public class ReturnStructFromCallbacksStack : Stack
    {
        public class MyEventTarget : Events.IRuleTarget {
          public Events.IRuleTargetConfig Bind(Events.IRule rule, string id) {
            return new Events.RuleTargetConfig() {
                Id = "asdf",
                Arn = "target-arn",
                EcsParameters = new Events.CfnRule.EcsParametersProperty() {
                  TaskDefinitionArn = "task-definition-arn"
                }
            };
          }
        }

        public class MyLbTarget : IApplicationLoadBalancerTarget
        {
            public ILoadBalancerTargetProps AttachToApplicationTargetGroup(IApplicationTargetGroup tg) {
                return new LoadBalancerTargetProps() {
                    TargetType = TargetType.IP,
                    TargetJson = new {
                        port = 1234,
                        id = "i-1234"
                    }
                };
            }
        }

        public class MyListenerProps : IBaseApplicationListenerProps
        {
            public double? Port { get; set; }
        }

        public class MyTargetProps : IAddApplicationTargetsProps
        {
            public double? Port { get; set; }
            public IApplicationLoadBalancerTarget[] Targets { get; set; }
        }

        public ReturnStructFromCallbacksStack(Construct scope, string id, IStackProps props) : base(scope, id, props)
        {
            var rule = new Events.Rule(this, "Rule", new Events.RuleProps() {
                Targets = new Events.IRuleTarget[] {
                    new MyEventTarget()
                }
            });

            var lb = new ApplicationLoadBalancer(this, "LB", new ApplicationLoadBalancerProps() {
                Vpc = new Vpc(this, "VPC")
            });
            var listener = lb.AddListener("Listener", new MyListenerProps() {
                Port = 80
            });
            listener.AddTargets("DefaultTargets", new MyTargetProps() {
                Port = 80,
                Targets = new IApplicationLoadBalancerTarget[] {
                    new MyLbTarget()
                }
            });
        }
    }
}

File: typescript/return-struct-from-callbacks/lib/return-struct-from-callbacks-stack.ts
import ec2 = require('@aws-cdk/aws-ec2');
import elb = require('@aws-cdk/aws-elasticloadbalancingv2');
import cdk = require('@aws-cdk/core');

export class ReturnStructFromCallbacksStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const lb = new elb.ApplicationLoadBalancer(this, 'LB', { vpc: new ec2.Vpc(this, 'Vpc') });
    const listener = lb.addListener('Listener', { port: 80 });
    listener.addTargets('DefaultTargets', { targets: [new MyLBTarget()], port: 80 });
  }
}

class MyLBTarget implements elb.IApplicationLoadBalancerTarget {
  public attachToApplicationTargetGroup(targetGroup: elb.IApplicationTargetGroup): elb.LoadBalancerTargetProps {
    return {
      targetType: elb.TargetType.IP,
      targetJson: { port: 1234, id: 'i-1234' }
    };
  }
}
